File: workers/spreadsheet/ledger_export.mjs
import fs from "node:fs/promises";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const outputPath = process.argv[2];
if (!outputPath) throw new Error("output path is required");
let stdin = "";
for await (const chunk of process.stdin) stdin += chunk;
const payload = JSON.parse(stdin);

const wallTime = (value, timeZone) => {
  const parts = Object.fromEntries(new Intl.DateTimeFormat("en-CA", {
    timeZone, year: "numeric", month: "2-digit", day: "2-digit", hour: "2-digit", minute: "2-digit", second: "2-digit", hourCycle: "h23",
  }).formatToParts(new Date(value)).filter((part) => part.type !== "literal").map((part) => [part.type, Number(part.value)]));
  return new Date(Date.UTC(parts.year, parts.month - 1, parts.day, parts.hour, parts.minute, parts.second));
};
const localLabel = (value, timeZone) => new Intl.DateTimeFormat("zh-CN", {
  timeZone, year: "numeric", month: "2-digit", day: "2-digit", hour: "2-digit", minute: "2-digit", hourCycle: "h23",
}).format(new Date(value)).replaceAll("/", "-");

const workbook = Workbook.create();
const detail = workbook.worksheets.add("账单明细");
const summary = workbook.worksheets.add("月度汇总");
detail.showGridLines = false;
summary.showGridLines = false;

detail.getRange("A1:H1").merge();
detail.getRange("A1").values = [[`伴AI 账单明细 · ${payload.month}`]];
detail.getRange("A1:H1").format = { fill: "#C65D2E", font: { bold: true, color: "#FFFFFF", size: 16 }, rowHeight: 32 };
detail.getRange("A2:H2").merge();
detail.getRange("A2").values = [[`生成时间：${localLabel(payload.generated_at, payload.timezone)} · 报表时区：${payload.timezone}`]];
detail.getRange("A2:H2").format = { fill: "#FFF3E8", font: { color: "#6B4A36", size: 10 }, rowHeight: 22 };
detail.getRange("A5:H5").values = [["发生时间", "收支方向", "分类", "商户", "备注", "币种", "金额", "记录 ID"]];
detail.getRange("A5:H5").format = { fill: "#6B7C5A", font: { bold: true, color: "#FFFFFF" }, rowHeight: 24 };
const rows = payload.entries.map((entry) => [wallTime(entry.occurred_at, payload.timezone), entry.direction, entry.category, entry.merchant, entry.note, entry.currency, entry.amount_minor / 100, entry.id]);
if (rows.length > 0) {
  detail.getRangeByIndexes(5, 0, rows.length, 8).values = rows;
  detail.getRange(`A6:A${5 + rows.length}`).format.numberFormat = "yyyy-mm-dd hh:mm";
  detail.getRange(`G6:G${5 + rows.length}`).format.numberFormat = "#,##0.00";
  detail.getRange(`A6:H${5 + rows.length}`).format.borders = { preset: "inside", style: "thin", color: "#E7DDD4" };
  detail.tables.add(`A5:H${5 + rows.length}`, true, "LedgerEntriesTable").style = "TableStyleMedium4";
}
detail.freezePanes.freezeRows(5);
detail.getRange("A:H").format.autofitColumns();
detail.getRange("A:A").format.columnWidth = 20;
detail.getRange("D:E").format.columnWidth = 18;
detail.getRange("H:H").format.columnWidth = 38;

summary.getRange("A1:F1").merge();
summary.getRange("A1").values = [[`伴AI 月度账本 · ${payload.month}`]];
summary.getRange("A1:F1").format = { fill: "#C65D2E", font: { bold: true, color: "#FFFFFF", size: 16 }, rowHeight: 32 };
summary.getRange("A2:B2").values = [["统计币种", payload.currency]];
summary.getRange("A4:A7").values = [["收入"], ["支出"], ["结余"], ["记录数"]];
summary.getRange("A4:A7").format = { fill: "#FFF3E8", font: { bold: true, color: "#6B4A36" } };
const dataEnd = Math.max(105, 5 + rows.length);
summary.getRange("B4").formulas = [[`=SUMIFS('账单明细'!$G$6:$G$${dataEnd},'账单明细'!$B$6:$B$${dataEnd},"income",'账单明细'!$F$6:$F$${dataEnd},$B$2)`]];
summary.getRange("B5").formulas = [[`=SUMIFS('账单明细'!$G$6:$G$${dataEnd},'账单明细'!$B$6:$B$${dataEnd},"expense",'账单明细'!$F$6:$F$${dataEnd},$B$2)`]];
summary.getRange("B6").formulas = [["=B4-B5"]];
summary.getRange("B7").formulas = [[`=COUNTIFS('账单明细'!$F$6:$F$${dataEnd},$B$2)`]];
summary.getRange("B4:B6").format.numberFormat = "#,##0.00";
summary.getRange("A4:B7").format.borders = { preset: "outside", style: "thin", color: "#D7C5B8" };
const categories = [...new Set(payload.entries.filter((entry) => entry.direction === "expense" && entry.currency === payload.currency).map((entry) => entry.category))].sort();
summary.getRange("A10:B10").values = [["支出分类", "金额"]];
summary.getRange("A10:B10").format = { fill: "#6B7C5A", font: { bold: true, color: "#FFFFFF" } };
categories.forEach((category, index) => {
  const row = 11 + index;
  summary.getRange(`A${row}`).values = [[category]];
  summary.getRange(`B${row}`).formulas = [[`=SUMIFS('账单明细'!$G$6:$G$${dataEnd},'账单明细'!$B$6:$B$${dataEnd},"expense",'账单明细'!$C$6:$C$${dataEnd},A${row},'账单明细'!$F$6:$F$${dataEnd},$B$2)`]];
});
if (categories.length > 0) {
  summary.getRange(`B11:B${10 + categories.length}`).format.numberFormat = "#,##0.00";
  const chart = summary.charts.add("bar", summary.getRange(`A10:B${10 + categories.length}`));
  chart.title = "支出分类分布";
  chart.hasLegend = false;
  chart.yAxis = { numberFormatCode: "#,##0.00" };
  chart.setPosition("D3", "K18");
}
summary.getRange("A:B").format.autofitColumns();
summary.getRange("A:A").format.columnWidth = 18;
summary.getRange("B:B").format.columnWidth = 16;
const xlsx = await SpreadsheetFile.exportXlsx(workbook);
await xlsx.save(outputPath);

File: workers/python/src/ai_companion_worker/ledger_export.py
"""Generate the monthly ledger workbook used by the Go ledger exporter."""

from __future__ import annotations

import json
import sys
from datetime import datetime
from pathlib import Path
from typing import Any
from zoneinfo import ZoneInfo

from openpyxl import Workbook
from openpyxl.chart import BarChart, Reference
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.worksheet.table import Table, TableStyleInfo


def _parse_time(value: str, timezone: str) -> datetime:
    instant = datetime.fromisoformat(value.replace("Z", "+00:00"))
    return instant.astimezone(ZoneInfo(timezone)).replace(tzinfo=None)


def export_workbook(payload: dict[str, Any], output_path: Path) -> None:
    month = str(payload["month"])
    currency = str(payload["currency"])
    timezone = str(payload["timezone"])
    entries = list(payload.get("entries", []))
    generated_at = _parse_time(str(payload["generated_at"]), timezone)

    workbook = Workbook()
    detail = workbook.active
    detail.title = "账单明细"
    summary = workbook.create_sheet("月度汇总")
    detail.sheet_view.showGridLines = False
    summary.sheet_view.showGridLines = False

    orange = "C65D2E"
    green = "6B7C5A"
    cream = "FFF3E8"
    white = "FFFFFF"
    brown = "6B4A36"
    thin = Side(style="thin", color="E7DDD4")

    detail.merge_cells("A1:H1")
    detail["A1"] = f"伴AI 账单明细 · {month}"
    detail["A1"].fill = PatternFill("solid", fgColor=orange)
    detail["A1"].font = Font(bold=True, color=white, size=16)
    detail["A1"].alignment = Alignment(vertical="center")
    detail.row_dimensions[1].height = 32
    detail.merge_cells("A2:H2")
    detail["A2"] = f"生成时间：{generated_at:%Y-%m-%d %H:%M} · 报表时区：{timezone}"
    detail["A2"].fill = PatternFill("solid", fgColor=cream)
    detail["A2"].font = Font(color=brown, size=10)
    detail.row_dimensions[2].height = 22

    headers = ["发生时间", "收支方向", "分类", "商户", "备注", "币种", "金额", "记录 ID"]
    for column, value in enumerate(headers, 1):
        cell = detail.cell(row=5, column=column, value=value)
        cell.fill = PatternFill("solid", fgColor=green)
        cell.font = Font(bold=True, color=white)
    detail.row_dimensions[5].height = 24

    for row_number, entry in enumerate(entries, 6):
        values = [
            _parse_time(str(entry["occurred_at"]), timezone),
            entry.get("direction", ""),
            entry.get("category", ""),
            entry.get("merchant", ""),
            entry.get("note", ""),
            entry.get("currency", ""),
            int(entry.get("amount_minor", 0)) / 100,
            entry.get("id", ""),
        ]
        for column, value in enumerate(values, 1):
            cell = detail.cell(row=row_number, column=column, value=value)
            cell.border = Border(left=thin, right=thin, top=thin, bottom=thin)
        detail.cell(row=row_number, column=1).number_format = "yyyy-mm-dd hh:mm"
        detail.cell(row=row_number, column=7).number_format = "#,##0.00"

    if entries:
        table = Table(displayName="LedgerEntriesTable", ref=f"A5:H{5 + len(entries)}")
        table.tableStyleInfo = TableStyleInfo(name="TableStyleMedium4", showRowStripes=True)
        detail.add_table(table)
    detail.freeze_panes = "A6"
    widths = {"A": 20, "B": 12, "C": 14, "D": 18, "E": 22, "F": 10, "G": 14, "H": 38}
    for column_letter, column_width in widths.items():
        detail.column_dimensions[column_letter].width = column_width

    summary.merge_cells("A1:F1")
    summary["A1"] = f"伴AI 月度账本 · {month}"
    summary["A1"].fill = PatternFill("solid", fgColor=orange)
    summary["A1"].font = Font(bold=True, color=white, size=16)
    summary.row_dimensions[1].height = 32
    summary["A2"] = "统计币种"
    summary["B2"] = currency
    labels = ["收入", "支出", "结余", "记录数"]
    for row, label in enumerate(labels, 4):
        summary.cell(row=row, column=1, value=label)
        summary.cell(row=row, column=1).fill = PatternFill("solid", fgColor=cream)
        summary.cell(row=row, column=1).font = Font(bold=True, color=brown)

    data_end = max(105, 5 + len(entries))
    summary["B4"] = f'=SUMIFS(\'账单明细\'!$G$6:$G${data_end},\'账单明细\'!$B$6:$B${data_end},"income",\'账单明细\'!$F$6:$F${data_end},$B$2)'
    summary["B5"] = f'=SUMIFS(\'账单明细\'!$G$6:$G${data_end},\'账单明细\'!$B$6:$B${data_end},"expense",\'账单明细\'!$F$6:$F${data_end},$B$2)'
    summary["B6"] = "=B4-B5"
    summary["B7"] = f'=COUNTIFS(\'账单明细\'!$F$6:$F${data_end},$B$2)'
    for row in range(4, 7):
        summary.cell(row=row, column=2).number_format = "#,##0.00"

    categories = sorted(
        {
            str(entry.get("category", ""))
            for entry in entries
            if entry.get("direction") == "expense" and entry.get("currency") == currency
        }
    )
    summary["A10"] = "支出分类"
    summary["B10"] = "金额"
    for cell in summary[10][:2]:
        cell.fill = PatternFill("solid", fgColor=green)
        cell.font = Font(bold=True, color=white)
    for index, category in enumerate(categories, 11):
        summary.cell(row=index, column=1, value=category)
        summary.cell(
            row=index,
            column=2,
            value=f'=SUMIFS(\'账单明细\'!$G$6:$G${data_end},\'账单明细\'!$B$6:$B${data_end},"expense",\'账单明细\'!$C$6:$C${data_end},A{index},\'账单明细\'!$F$6:$F${data_end},$B$2)',
        ).number_format = "#,##0.00"
    if categories:
        chart = BarChart()
        chart.title = "支出分类分布"
        chart.legend = None
        chart.add_data(Reference(summary, min_col=2, min_row=10, max_row=10 + len(categories)), titles_from_data=True)
        chart.set_categories(Reference(summary, min_col=1, min_row=11, max_row=10 + len(categories)))
        chart.height = 8
        chart.width = 14
        summary.add_chart(chart, "D3")
    summary.column_dimensions["A"].width = 18
    summary.column_dimensions["B"].width = 16

    output_path.parent.mkdir(parents=True, exist_ok=True)
    workbook.save(output_path)


def main() -> None:
    if len(sys.argv) != 2:
        raise SystemExit("output path is required")
    export_workbook(json.load(sys.stdin), Path(sys.argv[1]))


if __name__ == "__main__":
    main()
